design-proposals/orch-cli.md

```diff
@@ -0,0 +1,229 @@
+
+- The CLI should avoid distinguishing between singular and plural nouns. For example `list applications`
+  and `list application` should be equivalent commands.
+
+- The CLI stores its local state in the file (in a Linux environment) `~/.orch-cli/orch-cli.yaml`. This state
+  includes the orchestrator endpoint, keycloak endpoint, default projects, logged in username, and logged in
+  refresh token. This file should have permissions set appropriately.
+
+- The CLI generally uses the single `api.` endpoint for the orchestrator. There are occasionally places where
+  it may infer the names of additional endpoints. For example, during login it will infer the `keycloak.` endpoint
+  by taking the `api.` endpoint and substituting `keylocak` for `api`. This may be done with other endpoints
+  as necessary.
+
+## Rationale
+
+A fully autogenerated approach was discussed, but we feel this would too closely tie the CLI to the syntax
+and structure of the API. The CLI is meant to abstract the operations of the API into a more human-consumable
+format.
+
+- There may be times when a particular CLI operation leads to multiple API calls.
+
+- Related objects may be internally resolved by the CLI as necessary. For example, a user might link objects A
+  and B together by name, even if the API represents the link by an internal identifiers or uuids.
+
+- If there are future breaking changes to the API, then the CLI may seek to insulate the user from those changes.
+
+- Addtional context or explanation may be provided in the CLI, such as returning objects in a human-readable format.
+
+- The CLI may contain additional guardrails to "fail early" or to fail in a human readable manner when the user
+  performs an incorrect operation.
+
+- The CLI may seek to hide or mask sensitive information such as tokens or passwords.
+
+## Affected components and Teams
+
+Application Orchestration, Cluster Orchestration, EIM, Platform.
+
+## Implementation plan
+
+The CLI is to be implemented in the `go` programming language, using the popular `cobra` and `viper`
+go libraries.
+
+The CLI shall include a Makefile and github actions that facilitate building it for Linux, Windows, and
+Mac.
+
+### Code structure
+
+The CLI shall be implemented in a modular way.
+
+Each noun shall be placed in a file `<noun>.go` in the `internal/cli` directory. For example,
+
+- `internal/cli/application.go` ... implements the noun `application` and its associated verbs.
+
+TO-DO: Here is may make sense to separate code by subsystem for easier maintainability. For example,
+`internal/cli/catalog/application.go`, `internal/cli/cluster/cluster-templates.go`, etc. Discuss.
+
+## Open issues (if applicable)
+
+- Session / Context management. It would be convenient if the CLI could store the context for
+  multiple orchestrator sessions and then easily switch between them.
+
+- Distribution of Binaries. Proposal is to make them a github release, so they may be easily
```

Comment:
to me this is not an open issue but a fact.

Reply:
moved.